0015-estimator-interface.md

```diff
@@ -0,0 +1,457 @@
+
+1. _Performance._ Given that `QuantumCircuit` hashing should be avoided and circuit equality checks can be expensive, we should move away from an interface that necessitates performant primitive implementations to go down this path at all. As one example, qubit-wise commuting observables like `"IZZ"` and `"XZI"` can both be estimated using the same simulation/execution, but only if the estimator understands that they share a base circuit. As a second example, when the circuits need to be seriazed before they are simulated/executed (e.g. runtime or pickling-for-multi-processing), it puts the onus on the primitive implementation to detect circuit duplications.
+   
+
+Here is why the [Detailed Design](#detailed-design) section suggests "transposing" the signature and using array-based arguments with broadcasting:
+
+ 1. Transposing the signature will make it obvious what the primitive intends to do with each circuit.
+ 2. Transposing the signature will let us introduce and reason about the notion  "primitive unit of work", for which this RFC proposes the name "task".
+ 3. Array-based arguments will let users assign operational meaning to each axis (this axis is for twirling, that axis is for basis changes, etc.).
+ 4. Broadcasting rules will let users choose how to combine parameter value sets with observables in different ways, such as
+    1. use one observable for all of N parameter value sets
+    2. use one parameter value set for all of N observables
+    3. zip N parameter value sets against N observables
+    4. take all NxM combinations of N parameter value sets with M observables
+    5. etc.
+
+## User Benefit <a name="user-benefit"></a>
+
+All users of the primitives stand to benefit from this proposal.
+Immediately, it will enable sophisticated and convenient workflows for power users though arrays and broadcasting.
+However, standard broadcasting rules are such that the 0D and 1D cases will feel natural to existing users---0D is essentially what we already have, and 1D will be perceived as a simple-yet-welcome bonus.
+
+Moreover, the interface changes in this proposal will enable specific primitive implementations to enhance performance through reduced bandwidth on the runtime, compatibility with fast parameter binding, and multiplexing qubit-wise commuting observables.
+
+## Design Proposal <a name="design-proposal"></a>
+
+We will lead the design proposal with an example:
+
+```python
+# construct two circuits, the first with 2039 Parameters, the second with no Parameters
+circuit1 = QuantumCircuit(9)
+...
+circuit2 = QuantumCircuit(2)
+...
+
+# specify 128 different parameter value sets for circuit1, in a 4x32 shape
+parameter_values1 = np.random.uniform((4, 32, 2039))
```

Comment:
I don't think that's what this line does.
```python
>>> np.random.uniform((4, 32, 2039))
array([  2.26565035,   8.87576619, 446.18998942])
```

Reply:
Thanks, fixed in e400b2484de6a.